### ISSUE TITLE
What dependencies are required to install WarpX on a new HPC?

### ISSUE THREAD

**Issue by jcyu96** [ECP-WarpX/WarpX#3428]
Hi developers,
I'm installing WarpX on a new HPC, but can't dump data in `.h5` format. How should I compile WarpX properly? Besides, what further details do I need to offer to help solve this issue? 
Thanks very much!

**Comment by ax3l** [comment]
Hi @jcyu96,

Thanks for opening the issue.  I saw you also posted on our Gitter and will provide the details for your question here.

> I'm installing WarpX on a new HPC, but can't dump data in .h5 format. How should I compile WarpX properly?

So the main thing you want to look out for is the summary table of your first CMake call.
This might look like this:
```console
cmake -S . -B build
```
```
...
-- Could NOT find HDF5 (missing: HDF5_LIBRARIES HDF5_INCLUDE_DIRS C) (found version "")
...
openPMD build configuration:
  library Version: 0.14.5
  openPMD Standard: 1.1.0
  C++ Compiler: GNU 9.4.0 
    /usr/bin/c++

  Installation: OFF

  Build Type: Release
  Library: static
  CLI Tools: OFF
  Examples: OFF
  Testing: OFF
  Invasive Tests: OFF
  Internal VERIFY: ON
  Build Options:
    MPI: ON
    HDF5: OFF
    ADIOS1: OFF
    ADIOS2: OFF
    PYTHON: OFF

-- Found Git: /usr/bin/git (found version "2.25.1") 

WarpX build configuration:
  Version: 22.09 (22.09-50-g837574c52618)
  C++ Compiler: GNU 9.4.0 
    /usr/bin/c++

  Installation prefix: /usr/local
        bin: bin
        lib: lib
    include: include
      cmake: lib/cmake/WarpX

  Build type: Release
  Build options:
    APP: ON
    ASCENT: OFF
    COMPUTE: OMP
    DIMS: 3
    Embedded Boundary: OFF
    GPU clock timers: OFF
    IPO/LTO: OFF
    LIB: OFF
    MPI: ON
    PSATD: OFF
    PRECISION: DOUBLE
    PARTICLE PRECISION: DOUBLE
    OPENPMD: ON
    QED: ON
    QED table generation: OFF
    SENSEI: OFF

-- Configuring done
-- Generating done
-- Build files have been written to: /home/axel/src/warpx/build
```

In this case, the `HDF5: OFF` entry indicates there is no parallel HDF5 found. If you scroll up further, you will find more hints why this is the case, e.g., HDF5 was found but was only serial or HDF5 was found and is too old or HDF5 was simply not found at all.

In case you are sure there is a matching HDF5, e.g., because you installed in manually, then consider hinting its location via environment variables.
We recommend [this CMake tutorial](https://hsf-training.github.io/hsf-training-cmake-webpage/) in the [developer install docs](https://warpx.readthedocs.io/en/latest/install/cmake.html).
The standardized CMake dependency hint for `find_package` is [CMAKE_PREFIX_PATH](https://cmake.org/cmake/help/latest/envvar/CMAKE_PREFIX_PATH.html):
```bash
# on Unix/Linux: assuming you installed HDF5 into the prefix "$HOME/sw/hdf5"
export CMAKE_PREFIX_PATH=$HOME/sw/hdf5:${CMAKE_PREFIX_PATH}
```

> What dependencies are required to install WarpX on a new HPC?

Generally, these dependencies are needed:
https://warpx.readthedocs.io/en/latest/install/dependencies.html#install-dependencies

HDF5 is an optional dependency. We use HDF5 for [openPMD](https://www.openpmd.org) output via [openPMD-api](https://openpmd-api.readthedocs.io). Alternatively, you can also use [ADIOS2 with openPMD output](https://warpx.readthedocs.io/en/latest/dataanalysis/formats.html).

> Besides, what further details do I need to offer to help solve this issue?

Since your question is related to an HPC system, the following info would help to progress:
- Which HPC system are you on?
- Does this HPC system have a public homepage/documentation that we can look together on?
- Are you a user or an HPC support professional of the system?
  - In case you are a user, are you already in contact with your HPC support team to ask them for software modules for dependencies (or WarpX itself)?
- What's the size of jobs you want to run in terms of number of nodes and MPI processes? If you go very large, I would recommend to also try ADIOS2 alongside of HDF5, due to usually better performance.

**Comment by jcyu96** [comment]
Hi @ax3l,

I have recently installed WarpX successfully on new HPC and dumped the .h5 data in h5 following [Cori's (NERSC) instructions](https://warpx.readthedocs.io/en/latest/install/hpc/cori.html).

Following is a short summary of my installation process: 

```
module load boost/172-gcc-cjj-public3
module load mpi/openmpi/4.1.0-icc20-public3
module load fftw/3.3.8-mpi-public3
module load hdf5/1.10.4-parallel-icc20-public3

export PKG_CONFIG_PATH=$FFTW_DIR/pkgconfig:$PKG_CONFIG_PATH
export CMAKE_PREFIX_PATH=/path/to/c-blosc-1.12.1-install:$CMAKE_PREFIX_PATH
export CMAKE_PREFIX_PATH=/path/to/adios2-2.7.1-install:$CMAKE_PREFIX_PATH

# The following steps are done in the personal user directory
# c-blosc (I/O compression)
git clone -b v1.21.1 https://github.com/Blosc/c-blosc.git ./c-blosc
cmake -S ./c-blosc -B ./c-blosc-build -DBUILD_TESTS=OFF -DBUILD_BENCHMARKS=OFF -DDEACTIVATE_AVX2=OFF -DCMAKE_INSTALL_PREFIX=/path/to/c-blosc-1.12.1-install
cmake --build ./c-blosc-build --target install --parallel 16

# ADIOS2
git clone -b v2.7.1 https://github.com/ornladios/ADIOS2.git ./adios2
cmake -S ./adios2 -B src/adios2-build -DADIOS2_USE_Blosc=ON -DADIOS2_USE_Fortran=OFF -DADIOS2_USE_Python=OFF -DADIOS2_USE_ZeroMQ=OFF -DADIOS2_USE_HDF5=ON -DCMAKE_INSTALL_PREFIX=/path/to/adios2-2.7.1-install
cmake --build src/adios2-build --target install --parallel 16

cd /path/to/warpx
cmake -S . -B build -DWarpX_DIMS=2 
cmake --build build -j 16
```

Additionally,

- I'm running on a HPC from [BEIJING SUPER CLOUD COMPUTING CENTER](http://www.blsc.cn/index.asp).
- I haven't found Doc on its [website](http://www.blsc.cn/index.asp) or anywhere else.
- I'm a user of the system, and they had already installed the required module dependencies for me.
- For running my simulations with better performance, I hope I can get further help from you. This question can be discussed in more detail.

I would run larger simulations in the future, such as 4000*4000, 128 particles per cell, 64*(4-8) CPUs. 
Perhaps I need to provide more information to help you determine how to improve my simulation performance, but I'm not sure exactly what information I need and look forward to receiving your response.

**Comment by jcyu96** [comment]
By the way, I have one more doubt. I'm not sure if I need to install BLAS++ and LAPACK++, like the following: 

```
# BLAS++ (for PSATD+RZ)
git clone https://bitbucket.org/icl/blaspp.git src/blaspp
rm -rf src/blaspp-haswell-build
cmake -S src/blaspp -B src/blaspp-haswell-build -Duse_openmp=ON -Duse_cmake_find_blas=ON -DBLAS_LIBRARIES=${CRAY_LIBSCI_PREFIX_DIR}/lib/libsci_gnu.a -DCMAKE_CXX_STANDARD=17 -DCMAKE_INSTALL_PREFIX=$HOME/sw/blaspp-master-haswell-install
cmake --build src/blaspp-haswell-build --target install --parallel 16

# LAPACK++ (for PSATD+RZ)
git clone https://bitbucket.org/icl/lapackpp.git src/lapackpp
rm -rf src/lapackpp-haswell-build
CXXFLAGS="-DLAPACK_FORTRAN_ADD_" cmake -S src/lapackpp -B src/lapackpp-haswell-build -Duse_cmake_find_lapack=ON -DBLAS_LIBRARIES=${CRAY_LIBSCI_PREFIX_DIR}/lib/libsci_gnu.a -DLAPACK_LIBRARIES=${CRAY_LIBSCI_PREFIX_DIR}/lib/libsci_gnu.a -DCMAKE_CXX_STANDARD=17 -DCMAKE_INSTALL_PREFIX=$HOME/sw/haswell/lapackpp-master-install
cmake --build src/lapackpp-haswell-build --target install --parallel 16
```

This module seems to provide support for Pseudo Spectral Analytical Time Domain and RZ geometry. Can PSATD help enhance the performance of the simulation?

**Comment by ax3l** [comment]
@jcyu96 awesome, congrats on the progress!

Do you like to add a new entry for this HPC machine [in our docs](https://warpx.readthedocs.io/en/latest/install/hpc.html) via a pull request? You can take an existing machine as a template:
  https://github.com/ECP-WarpX/WarpX/tree/development/Docs/source/install/hpc

> For running my simulations with better performance, I hope I can get further help from you. This question can be discussed in more detail.

Sounds good - maybe open separate issues for these questions if they are becoming more specific for various cases.

For the current issue, let's also post your submission script template that you use and info on the nodes you run on - for CPU runs, pinning of MPI ranks and OpenMP threads is quite important for good performance and its done in that script.

> Can PSATD help enhance the performance of the simulation?

PSATD is [an advanced, electromagnetic field solver](https://warpx.readthedocs.io/en/latest/acknowledge_us.html#prior-warpx-references) that we are developing. It provides you with a virtually dispersion free field evolution (contrary to FDTD solvers) and can mitigate numerical instabilities in combination with further methods, i.e., for boosted frame simulations.

A PSATD step is more costly than a simpler FDTD stencil solve. But, since it helps you reduce numerical artifacts, you might be able to run a simulation with less resolution - which in turn for large 3D simulations can indeed result in saved time-to-solution.

So in summary: it depends if it lets you get results faster - possibly. Primarily it gives you results more precise and more stable, especially at long-running sims and at relativistically flowing plasmas/beams.

> BLAS++ & LAPACK++: Only needed for PSATD+RZ

BLAS++ and LAPACK++ are only needed if you want to run in quasi-cylindrical (Hankel-transformed) "RZ" geometry **with** the PSATD field solver. If your problem does not adhere to RZ geometry **or** if you don't need PSATD, then you don't need these two dependencies.

**Comment by jcyu96** [comment]
@ax3l , ok.

**Comment by jcyu96** [comment]
Okay @ax3l , and I' ll write an installation instruction for this HPC related later on.

**Comment by ax3l** [comment]
Awesome, thanks a lot! Feel free to open the PR in draft state @jcyu96 and we can finish it together in review.
